Add Initial Implementation of sage_sms Package for SMS Provider Integration

MohmdFo opened the pull request:
#### Summary
This pull request introduces the initial implementation of the `sage_sms` package, designed to facilitate the integration of different SMS providers. The package provides a structured and modular approach to managing SMS providers, ensuring easy extension and maintenance.

#### Impact
This initial implementation lays the groundwork for the `sage_sms` package, providing a clear and organized structure. Future contributions can build upon this foundation by adding specific SMS provider integrations, improving the utility and flexibility of the package.


#### Next Steps
- Add unit tests for core components.
- Enhance documentation.

Please review the changes and provide feedback. Your input is invaluable in refining this initial implementation and guiding future developments.

MohmdFo committed:
Initial commit

MohmdFo committed:
build(chore): start project python-sage-sms

Commit:
feat(sage_sms): add core structure for python-sage-sms package

- Added initial interfaces
- Implemented factory.py
- Set up base backend

MohmdFo committed:
refactor(BaseSMSBackend): load settings in init

MohmdFo committed:
feat(SmsIr): add smsir backend

MohmdFo committed:
build(chore): add toml package

MohmdFo committed:
feat(utils): add load_settings function

MohmdFo committed:
feat(SmsService): add SmsService to send sms

MohmdFo committed:
refactor(BackendModuleLoader): add discover_backends to load dynamicly backends

MohmdFo committed:
refactor(sage_sms): change backend dir to backends

MohmdFo committed:
refactor(SMSBackendFactory): Initialize it with settings and base package, refactor get_backend to instance method

- Added `__init__` method to `SMSBackendFactory` to initialize with `settings` and `base_package`.
- Called `BackendModuleLoader.discover_backends(base_package)` in the `__init__` method to automatically discover backends upon initialization.
- Refactored `get_backend` to be an instance method instead of a static method, allowing access to instance-specific settings.
- Ensured `get_backend` method uses the initialized settings to determine the appropriate backend to load.

BREAKING CHANGE:

MohmdFo committed:
refactor(BackendModuleLoader): use os in discover_backends

MohmdFo committed:
refactor(BackendModuleLoader): ensure load_backend_module return the class obj

MohmdFo committed:
style(sage_sms): apply black rules on sage sms

Commit:
feat(sage_sms): Add custom exceptions and types for SMS settings and provider configuration

Commit:
feat(validators): Add PhoneNumberDescriptor for phone number validation and formatting

- Introduced PhoneNumberDescriptor class to handle phone number validation and formatting using the `phonenumbers` library.
- Updated PhoneNumberValidator to utilize PhoneNumberDescriptor for phone number attribute.
- Ensures phone numbers are parsed and formatted to E.164 standard upon setting.
- Raises appropriate errors for unset phone numbers or if the `phonenumbers` library is not installed.

Commit:
refactor(Add-custom-types,-exceptions,-and-logging-to-SMS-backend-handling): - Introduced `ProviderSettings` and `SMSSettings` custom types for better type safety.
- Added custom exceptions `SMSBackendError`, `SMSConfigurationError`, `SMSProviderNotFoundError`, and `SMSUnexpectedError` for improved error handling.
- Implemented logging configuration and debug logs to trace backend discovery and loading processes.
- Updated `BackendModuleLoader` to cache backend modules and handle discovery and loading with detailed logging.
- Enhanced `SMSBackendFactory` to utilize the new logging and custom exceptions, ensuring robust backend initialization

MohmdFo committed:
style(sage_sms): change import formatting

- update CONTRIBUTORS and pyproject.toml structure

MohmdFo committed:
docs(sphinx): add sphinx documentation to how to use with sage-sms

MohmdFo committed:
docs(backends): enhance the description of index

- make short examples short and useful

MohmdFo committed:
build(packages): add requirements-dev.txt

- add poetry.md for how to export file containing both the main dependencies and the development

MohmdFo committed:
refactor(chore): add more ignore files to .gitignore

- add .readthedocs.yaml config and naming changes in setup.py

MohmdFo committed:
docs(README): Write a short and imperative summary of how to use python-sage-sms

MohmdFo committed:
Merge remote-tracking branch 'local-repo/main'

MohmdFo committed:
Merge remote-tracking branch 'local-repo/develop'

sepehr-akbarzadeh commented on sage_sms/factory.py:
```diff
@@ -0,0 +1,98 @@
+from sage_sms.helper.type import ProviderSettings, SMSSettings
+
+# Configure logging
+logging.basicConfig(level=logging.DEBUG)
```
remove this line!

sepehr-akbarzadeh commented on pyproject.toml:
```diff
@@ -0,0 +1,89 @@
+version = "0.1.0"
+description = "A Python package for sending SMS with different providers"
+authors = [
+    "Sepehr Akbarzadeh <sa.goldeneagle@gmail.com>",
```
fix email addresses. it must be organizational emails for members!

sepehr-akbarzadeh commented on mypy.ini:
remove this file!

sepehr-akbarzadeh commented on .pylintrc:
remove this file!

sepehr-akbarzadeh commented on .flake8:
remove this file!

sepehr-akbarzadeh commented on setup.cfg:
remove this file!

sepehr-akbarzadeh commented on setup.py:
remove this file

sepehr-akbarzadeh commented:
Thank you for your collaboration on this project.

I have reviewed your PR and **found 7 issues** that need to be addressed. Could you please fix these issues first and then provide the updated PR?

Thanks again for your hard work and dedication.

MohmdFo committed:
build(sage_sms): Update and streamline project configuration files

- Updated author emails in `pyproject.toml` to new domain
- Refined `[tool.black]`, `[tool.isort]`, and `[tool.pylint]` configurations in `pyproject.toml`
- Updated `[tool.commitizen]` settings in `pyproject.toml`
- Removed redundant configuration files: `.flake8`, `.pylintrc`, `mypy.ini`, `setup.cfg`, `setup.py`
- Modified `sage_sms/factory.py` to reflect configuration changes

These updates streamline project configuration and ensure consistent management across the codebase.

MohmdFo commented:
I have completed all the requests mentioned above. I'd appreciate it if you could review them again.
@sepehr-akbarzadeh
